### PR TITLE
Preparations for eliminating arithmetic subtyping

### PR DESCRIPTION
**Pull request by ajreynol**
Adds TypeNode::isArithmetic, NodeManager::mkConstReal and NodeManager::mkConstInt.

The first method (for now) is equivalent to TypeNode::isReal, and the latter 2 are equivalent to NodeManager::mkConst(CONST_RATIONAL, ...).

This furthermore distinguishes all uses of `isReal`: all that intend to be `isArithmetic` are changed in this PR.  Redundant uses of `isReal() || isInteger()` are merged to `isArithmetic()`.

Due to the above, there are no behavior changes in this PR.

### CODE REVIEW

File: src/theory/quantifiers/term_util.cpp [hunk 1]
```diff
@@ -557,7 +557,7 @@ Node TermUtil::isSingularArg(Node n, Kind ik, unsigned arg)
   }
   else
   {
-    if (n.getType().isReal() && n.getConst<Rational>().sgn() < 0)
+    if (n.getType().isArithmetic() && n.getConst<Rational>().sgn() < 0)
```
Comment (4tXJ7f) [inline src/theory/quantifiers/term_util.cpp:560]: This should eventually become `n.getType().isInteger()`, no?

File: src/theory/quantifiers/sygus_sampler.cpp [hunk 2]
```diff
@@ -617,7 +617,7 @@ Node SygusSampler::getRandomValue(TypeNode tn)
       return ret;
     }
   }
-  else if (tn.isReal())
+  else if (tn.isArithmetic())
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus_sampler.cpp:620]: Shouldn't this remain `tn.isReal()`? The integer case is handled above.

File: src/theory/quantifiers/cegqi/ceg_arith_instantiator.cpp [hunk 3]
```diff
@@ -191,7 +191,7 @@ bool ArithInstantiator::processAssertion(CegInstantiator* ci,
         }
         else
         {
-          Assert(d_type.isReal());
+          Assert(d_type.isArithmetic());
```
Comment (4tXJ7f) [inline src/theory/quantifiers/cegqi/ceg_arith_instantiator.cpp:194]:
```suggestion
          Assert(d_type.isReal());
```
The integer case is handled above.

File: src/theory/quantifiers/cegqi/ceg_arith_instantiator.cpp [hunk 4]
```diff
@@ -262,7 +262,7 @@ bool ArithInstantiator::processAssertion(CegInstantiator* ci,
       }
       else
       {
-        Assert(d_type.isReal());
+        Assert(d_type.isArithmetic());
```
Comment (4tXJ7f) [inline src/theory/quantifiers/cegqi/ceg_arith_instantiator.cpp:265]:
```suggestion
        Assert(d_type.isReal());
```

The integer case is handled above.

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Some minor comments and we should consider changing the name of `isArithmetic()` as discussed during the meeting today.

File: src/theory/quantifiers/term_util.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/term_util.cpp:560]: True, I think we can change it now.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thanks for the updates.

**Pull request merged**